File: Chapter05/Lesson3/Exercise1/Completed/VB/Serialize-People/Person.vb
Imports System.Runtime.Serialization

<Serializable()> Class Person
    Implements ISerializable

    Public name As String
    Public dateOfBirth As DateTime
    Public age As Integer

    Public Sub New(ByVal _name As String, ByVal _dateOfBirth As DateTime)
        MyBase.New()
        name = _name
        dateOfBirth = _dateOfBirth
        CalculateAge()
    End Sub

    Public Sub New()
        MyBase.New()

    End Sub

    Public Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)
        name = info.GetString("Name")
        dateOfBirth = info.GetDateTime("DOB")
        CalculateAge()
    End Sub

    Public Overridable Sub GetObjectData(ByVal info As SerializationInfo, ByVal context As StreamingContext) Implements System.Runtime.Serialization.ISerializable.GetObjectData
        info.AddValue("Name", name)
        info.AddValue("DOB", dateOfBirth)
    End Sub

    Public Overrides Function ToString() As String
        Return (name + (" was born on " _
                    + (dateOfBirth.ToShortDateString + (" and is " _
                    + (age.ToString + " years old.")))))
    End Function

    Private Sub CalculateAge()
        age = (DateTime.Now.Year - dateOfBirth.Year)

        ' If they haven't had their birthday this year, 
        ' subtract a year from their age
        If (dateOfBirth.AddYears((DateTime.Now.Year - dateOfBirth.Year)) > DateTime.Now) Then
            age = (age - 1)
        End If
    End Sub

End Class

File: Chapter05/Lesson3/Exercise1/Completed/CS/Serialize-People/Person.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Runtime.Serialization;

namespace Serialize_People
{
    [Serializable]
    class Person : ISerializable
    {
        public string name;
        public DateTime dateOfBirth;
        public int age;

        public Person(string _name, DateTime _dateOfBirth)
        {
            name = _name;
            dateOfBirth = _dateOfBirth;
            CalculateAge();
        }

        public Person(SerializationInfo info, StreamingContext context)
        {
            name = info.GetString("Name");
            dateOfBirth = info.GetDateTime("DOB");
            CalculateAge();
        }

        // The following method is called during serialization
        public virtual void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            info.AddValue("Name", name);
            info.AddValue("DOB", dateOfBirth);
        }

        public Person()
        {
        }

        public override string ToString()
        {
            return name + " was born on " + dateOfBirth.ToShortDateString() + " and is " + age.ToString() + " years old.";
        }

        private void CalculateAge()
        {
            age = DateTime.Now.Year - dateOfBirth.Year;

            // If they haven't had their birthday this year, 
            // subtract a year from their age
            if (dateOfBirth.AddYears(DateTime.Now.Year - dateOfBirth.Year) > DateTime.Now)
            {
                age--;
            }
        }
    }
}
